src/Symfony/Component/DependencyInjection/ReverseContainer.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException;
+
+/**
+ * Turns public and "container.reversible" services back to their ids.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class ReverseContainer
+{
+    private $serviceContainer;
+    private $reversibleLocator;
+    private $getServiceId;
+
+    public function __construct(Container $serviceContainer, ContainerInterface $reversibleLocator, string $tagName = 'container.reversible')
+    {
+        $this->serviceContainer = $serviceContainer;
+        $this->reversibleLocator = $reversibleLocator;
+        $this->getServiceId = \Closure::bind(function ($service) {
+            return array_search($service, $this->services, true) ?: array_search($service, $this->privates, true);
+        }, $serviceContainer, Container::class);
+    }
+
+    /**
+     * @param object $service
+     *
+     * @throws ServiceNotFoundException When the service is not reversible and $throw is true
+     */
+    public function getId($service, bool $throw = false): ?string
```

Comment:
What's the design choice behind the throw option?

Reply:
It would allow building a generic normalizer for the Serializer component without paying the tax of building an exception for non service objects.

Reply:
I'm trying to think of a use-case where you'd want an exception instead. What if this function is renamed to `findId` instead and never throws an exception?

It does reduce 2 different explicitly defined error paths to `null` though :thinking:

Reply:
OK :)
That's because I think unhandled failure situations should throw by default. To make it "handled", pass true explicitly.

Reply:
then, should the signature be `bool $throw = true` instead ?

Reply:
true!